### PR TITLE
[close #138]  br return backup ts

### LINKED ISSUE #138: br: Return the timestamp of backup to cooperate with TiKV-CDC
## Feature Request

**Is your feature request related to a problem? Please describe:**

TiKV-BR is designed to initialize data before the incremental data replication of TiKV-CDC. Refer to [RawKV Cross Cluster Replication](https://github.com/pingyu/tikv-rfcs/blob/rawkv-cdc/text/0083-rawkv-cross-cluster-replication.md#1-utilize-tikv-br--tikv-cdc-as-data-replication-component) (https://github.com/tikv/rfcs/pull/86).

To cooperate with TiKV-CDC, the procedure is expected to be:
1. TiKV-BR backup from upstream TiKV
  1.1 **Get a latest TSO (supposed to be `T0`).**
  1.2 **Set BR service safepoint to `T0` and TTL as 72 hours (configurable).**
  1.3 **Invoke backup procedure to TiKV, flush `causal_ts` at the beginning to make sure that timestamp of following writes must be greater than `T0`.**
  1.4 **Return the `T0` after backup finished.**
2. TiKV-BR restore to downstream TiKV (It would be nice the store `T0` in backup meta to remind the users agin after restore).
3. TiKV-CDC create changefeed between upstream & downstream with `--start-ts T0`.

### Update 2022.6.29
Set `T0` as `TSO - safe-interval`. Where `safe-interval` is a configurable duration defaults to 1 minute.

`safe-interval` is a work around to avoid missing some RawKV writes just in Raft procedure during backup, and assume that all Raw writes should be finished in 1 minute.
The cons of `safe-interval` is that some entries would be backup or replicated redundantly.
On a system with heavy load, the `safe-interval` should be configured longer.
The `safe-interval` would be deprecated after the `read-ts` of RawKV is implemented. Refer to https://github.com/tikv/rfcs/pull/96.

**Describe the feature you'd like:**

1.1 ~ 1.4 in above description.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by haojinming**
### What problem does this PR solve?

Issue Number: close #138 

Problem Description: **TBD**

### What is changed and how does it work?
BR return the backup-ts, which is  ts described in issue #138 .

### Code changes

- Has exported function/method change

### Check List for Tests

This PR has been tested by at least one of the following methods:
- Unit test
- Manual test (add detailed scripts or steps below)
image

image

### Side effects

- No side effects

### CODE REVIEW

File: br/pkg/task/backup_raw.go [hunk 1]
```diff
@@ -49,6 +51,10 @@ func DefineRawBackupFlags(command *cobra.Command) {
 	command.Flags().Bool(flagRemoveSchedulers, false,
 		"disable the balance, shuffle and region-merge schedulers in PD to speed up backup.")
 
+	command.Flags().Duration(flagSafeInterval, utils.DefaultBRSafeInterval,
+		"The interval between backup-ts and current tso.")
+	command.Flags().Int64(flagGCTTL, utils.DefaultBRGCSafePointTTL, "the TTL (in seconds) that PD holds for BR's GC safepoint")
```
Comment (pingyu) [inline br/pkg/task/backup_raw.go:56]: Suggest to be a `duration` argument.
TTL would be expected to be e.g, 30 minutes, or 3 days. It is too big to be in secondes.

```suggestion
    command.Flags().Int64(flagGCTTL, utils.DefaultBRGCSafePointTTL, "the TTL of BR's GC safepoint")
```

File: br/pkg/task/backup_raw.go [hunk 2]
```diff
@@ -49,6 +51,10 @@ func DefineRawBackupFlags(command *cobra.Command) {
 	command.Flags().Bool(flagRemoveSchedulers, false,
 		"disable the balance, shuffle and region-merge schedulers in PD to speed up backup.")
 
+	command.Flags().Duration(flagSafeInterval, utils.DefaultBRSafeInterval,
```
Comment (pingyu) [inline br/pkg/task/backup_raw.go:54]: Suggest to hide safe interval. It's difficult for common users to set this argument.

File: br/pkg/utils/safe_point.go [hunk 3]
```diff
@@ -21,8 +21,9 @@ const (
 	brServiceSafePointIDFormat      = "br-%s"
 	preUpdateServiceSafePointFactor = 3
 	checkGCSafePointGapTime         = 5 * time.Second
-	// DefaultBRGCSafePointTTL means PD keep safePoint limit at least 5min.
-	DefaultBRGCSafePointTTL = 5 * 60
+	// DefaultBRGCSafePointTTL means PD keep safePoint limit at least 72h.
+	DefaultBRGCSafePointTTL = 72 * 60 * 60 // 72h
```
Comment (pingyu) [inline br/pkg/utils/safe_point.go:25]: Defaults to 72 hours may be too long, and waste disk space if there is no following CDC.
Maybe we can keep 5 minutes, but tell CDC users to set a suitable value.

**Review by pingyu — COMMENTED** [review]
Suggest to add related integration test, which extracts and verifies `backup-ts` from result of backup.

**Review by zeminzhou — APPROVED** [review]
LGTM~

File: br/pkg/backup/client_test.go [hunk 4]
```diff
@@ -52,7 +54,7 @@ func (r *testBackup) SetUpSuite(c *C) {
 	defer httpmock.DeactivateAndReset()
 	// Exact URL match
 	httpmock.RegisterResponder("GET", `=~^/config`,
-		httpmock.NewStringResponder(200, `{"storage":{"api-version":1, "enable-ttl":false}}`))
+		httpmock.NewStringResponder(200, `{"storage":{"api-version":2, "enable-ttl":false}}`))
```
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/backup/client_test.go]: done

File: br/pkg/task/backup_raw.go (hunk 1 above)
Comment (haojinming) [reply br/pkg/task/backup_raw.go]: done

File: br/pkg/task/backup_raw.go (hunk 2 above)
Comment (haojinming) [reply br/pkg/task/backup_raw.go:54]: done

File: br/pkg/utils/safe_point.go (hunk 3 above)
Comment (haojinming) [reply br/pkg/utils/safe_point.go]: done

**Comment by haojinming** [comment]
> Suggest to add related integration test, which extracts and verifies `backup-ts` from result of backup.

I would like to add integration test for apiv2 and `backup-ts` in next pr. May need another new github action for `apiv2` and need some more time to let it run correctly.

**Review by pingyu — APPROVED** [review]
LGTM~

**Pull request merged**